### PR TITLE
Add `—nilinit` for `redundantNilInit` rule

### PR DESCRIPTION
**Pull request by rakuyoMo**
Related https://github.com/nicklockwood/SwiftFormat/issues/1677 .

Added `—nilinit` option to `redundantNilInit` rule, its value includes `insert` and `remove`. Among them, `remove` is consistent with the previous logic, and `insert` is a new logic. It will add `nil` as the default value after the optional value type according to the opposite rules of `remove`.

> I did not implement the `properties-only` option mentioned in the issue, firstly because I did not need it, and secondly because I was not familiar enough with this project.

I added documentation and test cases and all of them passed.

This is my first PR for this project. If I did something wrong or if there is anything else that needs to be modified, please let me know.

### CODE REVIEW

File: Tests/RulesTests+Redundancy.swift [hunk 1]
```diff
@@ -1911,26 +1941,242 @@ class RedundancyTests: RulesTests {
...
+                       options: options)
     }
 
     func testNoRemoveNilInitInIfStatementInViewBuilder() {
         let input = """
         struct TestView: View {
             var body: some View {
                 if true {
-                    var foo: String?
+                    var foo: String? = nil
```
Comment (rakuyoMo) [inline Tests/RulesTests+Redundancy.swift:1954]: Did I miss adding nil as a default value here? If I understood it wrong, I will change it back, the same below.

File: Tests/RulesTests+Redundancy.swift [hunk 2]
```diff
@@ -1911,26 +1941,242 @@ class RedundancyTests: RulesTests {
...
     }
 
     func testNoRemoveNilInitInSwitchStatementInViewBuilder() {
+        let input = """
+        struct TestView: View {
+            var body: some View {
+                switch foo {
+                case .bar:
+                    var foo: String? = nil
+                    Text(foo ?? "")
```
Comment (rakuyoMo) [inline Tests/RulesTests+Redundancy.swift:1974]: Did I miss adding nil as a default value here? If I understand it wrong, I will change it back, same as above.

**Comment by rakuyoMo** [comment]
I fixed some issues and can now start the review.

**Comment by nicklockwood** [comment]
Thank you - this looks excellent. One thing I didn't previously consider is the case where the value is already being initialized in the init method, e.g.

```swift
class Foo {
    var foo: Int?// probably don't want to add = nil here?

    init() {
        foo = nil
    }
}
```

I checked and I don't think setting the value to nil twice will actually cause any compilation problems or changes in behavior, so it's probably fine to consider handling this an enhancement rather than a blocker for the feature though.

**Comment by rakuyoMo** [comment]
>  One thing I didn't previously consider is the case where the value is already being initialized in the init method.

If there is a rule in the future (I'm not sure if there is one now): "All properties that do not rely on external input should have their initial values set when they are defined, rather than set in the init method", then this problem will become simpler.

In my opinion, the correct solution to the situation you proposed is as follows:

```diff
class Foo {
-  var foo: Int?
+  var foo: Int? = nil

  // If init is empty, the entire init should also be removed. 
  // But I remember that there seems to be another rule that can automatically remove empty methods. 
  // I'm not sure how SwiftFormat handles the same parts between the two rules, is it implemented twice?
  init() {
-    foo = nil
  }
}
```

On this basis `properties-only` will be valuable, because users can use this option to implement "`insert` only applies to properties" (or implement another option similar to `variable-only` to only modify temporary variables , or both options).

I'm not handling the assignment in the init method at the moment, I thought I should at least implement it so that the `insert` option works perfectly in my vision.

-------

After I implement the above logic, `properties-only` or `variable-only` are still not needed for me. I'm sorry that I may not have the time or energy to implement it, unless you tell me that these two options will be necessary conditions for `insert`, or you have any other suggestions for how I can implement the above.

**Comment by rakuyoMo** [comment]
Please let me list other situations:

```diff
// --nilinit insert

class Foo {
-  var foo: Int?
+  var foo: Int? = nil

  init() {
-    foo = nil

-    var foo: String?
+    var foo: String? = nil
  }
}
```

```diff
// --nilinit insert-only-properties

class Foo {
-  var foo: Int?
+  var foo: Int? = nil

  init() {
-    foo = nil

    var foo: String? // nil will not be automatically inserted
  }
}
```

```diff
// --nilinit insert-only-variables

class Foo {
  var foo: Int? // nil will not be automatically inserted

  init() {
    foo = nil

-    var foo: String?
+    var foo: String? = nil
  }
}
```

**Comment by nicklockwood** [comment]
@rakuyoMo thanks for the update - I was not actually expecting you to fix this!

> After I implement the above logic, properties-only or variables-only are still not needed for me. I'm sorry that I may not have the time or energy to implement it, unless you tell me that these two options will be necessary conditions for insert, or you have any other suggestions for how I can implement the above.

This is absolutely fine not to include.

**Pull request merged**